Products/Zuul/catalog/model_catalog_init.py

```diff
@@ -109,6 +115,27 @@ def process(self, uid):
         """
         raise NotImplementedError
 
+    def acquire_semaphore(self):
+        if not self.semaphore_acquired:
+            self.semaphore.acquire(True)
+            self.semaphore_acquired = True
+
+    def release_semaphore(self):
+        if self.semaphore_acquired:
+            self.semaphore.release()
+            self.semaphore_acquired = False
+
+    def get_work_from_parent(self):
+        try:
+            # We hold the semaphore while we are waiting on the queue
+            # This is safe to call multiple times
+            self.acquire_semaphore()
+            new_work = self.parent_queue.get(timeout=QUEUE_TIMEOUT)
```

Comment:
It is intentional that the worker will continue to hold the semaphore until it succeeds in getting an item off the parent queue.  The parent process now terminates when all workers are holding the semaphore (vs. the opposite, as it was previously).  Note that calling acquire_semaphore will not actually do anything if we already have it.